extensions/Worker.Extensions.Http.AspNetCore/src/Coordinator/DefaultHttpCoordinator.cs

```diff
@@ -0,0 +1,66 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.Azure.Functions.Worker.Extensions.Http.AspNetCore.Coordinator
+{
+    internal class DefaultHttpCoordinator : IHttpCoordinator
+    {
+        private readonly ConcurrentDictionary<string, ContextReference> _contextReferenceList;
+
+        public DefaultHttpCoordinator()
+        {
+            _contextReferenceList = new ConcurrentDictionary<string, ContextReference>();
+        }
+
+        public Task<FunctionContext> SetHttpContextAsync(string invocationId, HttpContext context)
+        {
+            var contextRef = _contextReferenceList.GetOrAdd(invocationId, static id => new ContextReference(id));
```

Comment:
Simplified these to be GetOrAdd() and to use a static factory. AddOrUpdate() has unique race conditions that meant we couldn't be guaranteed of success with the way we were doing it (see Remarks): https://learn.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.addorupdate?view=net-7.0